File: OpenTheLock.java
class Solution {
    public int openLock(String[] deadends, String target) {
        HashSet<String> visited = new HashSet<>();
        Queue<String> red = new LinkedList<>();
        
        for(var d : deadends) {
            visited.add(d);
        }
        
        if(visited.contains("0000")) {
            return -1;
        }
        
        visited.add("0000");
        red.offer("0000");
        int steps = 0;
        
        while(!red.isEmpty()) {
            int len = red.size();
            for(int i=0; i<len; i++) {
                
                StringBuilder current = new StringBuilder(red.poll());
                if((current.toString()).equals(target)) {
                    return steps;
                }
                for(int j=0; j<4; j++) {
                    int inc = current.charAt(j) - '0';
                    int dec = current.charAt(j) - '0';

                    
                    inc = (inc + 1) % 10;
                    dec = (dec - 1 + 10) % 10;
                    StringBuilder comboInc = new StringBuilder(current);
                    StringBuilder comboDec = new StringBuilder(current);
                    
                    comboInc.setCharAt(j, (char) (inc + '0'));
                    comboDec.setCharAt(j, (char) (dec + '0'));
                    
                    if(!visited.contains(comboInc.toString())) {
                        visited.add(comboInc.toString());
                        red.offer(comboInc.toString());
                    }
                    if(!visited.contains(comboDec.toString())) {
                        visited.add(comboDec.toString());
                        red.offer(comboDec.toString());
                    }
                }
                
            }
            steps++;
        }
            
        
        return -1;
    }
}

File: OpenTheLock.cpp
class Solution {
public:
    int openLock(vector<string>& deadends, string target) {
        std::unordered_set<std::string> visited(deadends.begin(), deadends.end());
        // std::unordered_set<std::string> visited;
        std::queue<std::string> red;
        if(visited.count("0000") > 0) {
            return -1;
        }

        int steps{};
        red.push("0000");
        visited.insert("0000");
        while(!red.empty()) {
            int len = red.size();
            for(int i=0; i<len; i++) {
                if(red.front() == target) {
                    return steps;
                }
                for(int m=0; m<4; m++) {
                    int incDec[] = {1, -1};
                    for(int k=0; k<2; k++) {
                        std::string s = red.front();
                        int num = (static_cast<int>(s[m]) - '0') + incDec[k];
                        if(num == 10) {
                            num = 0;
                        } else if(num == -1) {
                            num = 9;
                        }
                        s[m] = '0' + num;
                        if(visited.count(s) == 0) {
                            red.push(s);
                            visited.insert(s);
                        }
                    }
                }
                red.pop();
            }
            steps++;
        }
        return -1;
    }
};
